## Обзор и анализ базы данных сервиса для чтения книг

**Цель проекта - для формулирования ценностного предложения проанализировать базу данных сервиса для чтения книг по подписке.**

`Задачи:`

- Посмотреть описание данных каждой таблицы и оценить данные;
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц (исключим из анализа брошюры);
- Определить автора с самой высокой средней оценкой книг (будем учитывать только книги с 50 и более оценками);
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

`Описание данных`:

*База данных содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг.*

Таблица **books** - cодержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors** - cодержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers** - cодержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица **ratings** - cодержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews** - cодержит данные о пользовательских обзорах на книги:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
= text — текст обзора.

## Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': _____________, # имя пользователя
'pwd': _______________, # пароль
'host': ___________________________,
'port':__________, # порт подключения
'db': _______________________} # название базы данных
connection_string = _______________.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={______________})

## Обзор и описание данных

In [3]:
def picture(sql):
    '''
    Обзорная функция для просмотра таблиц
    '''
    return pd.read_sql(sql,con = engine).info(), display(pd.read_sql(sql,con = engine).head())

In [4]:
def answer(sql):
    '''
    функция для запросов
    '''
    return pd.read_sql(sql,con = engine)

In [5]:
# составляем список похожих запросов для обработки их в цикле
sql = [
    '''SELECT*FROM books;''',
    '''SELECT*FROM authors;''',
    '''SELECT*FROM publishers;''',
    '''SELECT*FROM ratings;''',
    '''SELECT*FROM reviews;'''   
]

In [6]:
# вызываем функцию для наших данных
for i in sql:
    picture(i)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

- **таблица books** содержит информацию о 1000 книг, пропусков в данных нет
- **таблица authors** содержит имена 636 авторов, пропусков нет
- **таблица publishers** содержит названия 340 издательств, пропусков в данных нет
- **таблица ratings** содержит 6456 оценок книг, пропусков нет
- **таблица reviews** содержит 2793 обзора книг, пропусков в данных нет

## Cколько книг вышло после 1 января 2000 года?

Сначала посмотрим за какой временной период вообще есть книги в базе данных

In [7]:
# запрос на максимальную и минимальную даты выпуска книг
sql = '''
            SELECT MAX(publication_date),
                   MIN(publication_date)
            FROM books
        '''


answer(sql)

max         min
0  2020-03-31  1952-12-01

In [8]:
# узнаем сколько книг вышло после 1 января 2000 года
sql = '''
            SELECT COUNT (book_id)
            FROM books
            WHERE publication_date > '2000-01-01'

        '''


answer(sql)

count
0    819

**После 1 января 2000 года вышло 819 книг.**

### Ответ
У нас достаточно обширная по времени издания база данных, в ней книги, выпущенные с 1952 по 2020 год. После 1 января 2000 года вышло *819 книг*.

## Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [9]:
# формируем запрос
sql = '''WITH r_rw_table as
            (SELECT b.book_id,
                   COUNT (DISTINCT(rW.review_id)) AS count_review,
                   ROUND(AVG (r.rating),2) AS avg_rating
            FROM books b
            LEFT JOIN ratings r ON b.book_id=r.book_id
            LEFT JOIN reviews rw On b.book_id=rw.book_id
            GROUP BY b.book_id)
        SElECT b.book_id,
               b.title,
               a.author,
               r_rw.count_review,
               r_rw.avg_rating
        FROM books b
        JOIN r_rw_table  r_rw ON b.book_id=r_rw.book_id
        JOIN authors a ON b.author_id=a.author_id
        ORDER BY r_rw.count_review DESC, r_rw.avg_rating DESC
      '''

answer(sql)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...   
2        299  Harry Potter and the Chamber of Secrets (Harry...   
3        656                                     The Book Thief   
4        734                                   The Glass Castle   
..       ...                                                ...   
995      191  Disney's Beauty and the Beast (A Little Golden...   
996      387                               Leonardo's Notebooks   
997      221                          Essential Tales and Poems   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999      808                            The Natural Way to Draw   

                                                author  count_review  \
0                                      Stephenie Meyer             7   
1                           J.K. Rowling/Mary GrandPré             6   
2                           J.K. Rowling/Mary GrandPré             6   
3                    Markus Zusak/Cao Xuân Việt Khương             6   
4                                      Jeannette Walls             6   
..                                                 ...           ...   
995                 Teddy Slater/Ron Dias/Ric González             0   
996                      Leonardo da Vinci/H. Anna Suh             0   
997                 Edgar Allan Poe/Benjamin F. Fisher             0   
998  Faye Perozich/Anne Rice/John Bolton/Daerick Gröss             0   
999                      Kimon Nicolaides/Mamie Harmon             0   

     avg_rating  
0          3.66  
1          4.41  
2          4.29  
3          4.26  
4          4.21  
..          ...  
995        4.00  
996        4.00  
997        4.00  
998        3.67  
999        3.00  

[1000 rows x 5 columns]

### Ответ

Для каждой из 1000 книг мы посчитали количество обзоров и средний рейтинг. Максимальное количество обзоров 7, есть книги на которые нет обзоров, но рейтинговые данные есть у всех книг. Разными режимами сортировки мы выяснили, что максимальный средний рейстинг 5, минимальный 1.5. В итоговом выводе оставили сортировку в следующем порядке: 1. сначала по количеству обзоров, 2. затем по средней оценке.

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (так исключаем брошюры)

In [10]:
# формируем запрос
sql = '''
           SELECT p.publisher,
                  COUNT(b.book_id) AS count_book
           FROM books b
           JOIN publishers p ON b.publisher_id=p.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher_id
           ORDER BY count_book DESC
           LIMIT 1
            
        '''


answer(sql)

publisher  count_book
0  Penguin Books          42

### Ответ
- **Издательство, которое выпустило наибольшее число книг толще 50 страниц - *'Penguin Books'***, они выпустили 42 такие книги.

Предпросмотр результатов запроса показал, что:
- На втором месте издательство Vintage с 31 книгой толще 50 страниц,
- на третьем Grand Central Publishing	- 25 книг

## Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками

In [11]:
# формируем запрос
sql = '''SELECT a.author,
                ROUND(AVG(r.rating),1)
         FROM (SELECT b.book_id,
               AVG(r.rating) AS rating,
               COUNT(r.rating)
               FROM books AS b
               LEFT JOIN ratings AS r ON b.book_id = r.book_id
               GROUP BY b.book_id
               HAVING COUNT(r.rating) >= 50) AS r
        LEFT JOIN books AS b ON r.book_id = b.book_id
        LEFT JOIN authors AS a ON a.author_id = b.author_id
        GROUP BY a.author
        LIMIT 1
        '''


answer(sql)

author  round
0  J.K. Rowling/Mary GrandPré    4.3

### Ответ
**Лучше всего оценены книги:**
- **J.K. Rowling/Mary GrandPré (истории о мальчике, который выжил - Гарри Поттере)- средняя оценка книг 4.3 (при максимуме в 5 баллов)**

Предпросмотр результатов запроса показал, что далее сразу три автора получили среднюю оценку книг 3.8:
- Dan Brown (мистические/исторические детективы),
- J.D. Salinger	(американская классика с хитом "Над пропастью во ржи")
- William Shakespeare/Paul Werstine/Barbara A. M...	 (Шекспир и его английская классика)

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
# формируем запрос
sql = '''
            SELECT ROUND(AVG(x.avg), 0)
            FROM (SELECT COUNT (review_id) AS avg
            FROM reviews
            WHERE username IN (SELECT username                   
                              FROM ratings
                              GROUP BY username
                              HAVING COUNT(rating) > 50
                              ORDER BY COUNT(rating) DESC)
            GROUP BY username) AS x
        
        '''


answer(sql)

round
0   24.0

### Ответ
Наши активные пользователи (критерий: поставили более 50 оценок) пишут в среднем 24 обзора.

## Итоги

- У нас большая база данных: 1000 книг, 636 авторов, 340 издательств, работающая рейтинговая система - 6456 оценок книг, рабочий  функционал по написанию пользователями обзоров на прочитанные книги - 2793 обзора книг.
- Данные в базе качественные, пропусков нет
- Система рейтинговой оценки пятибальная, среди книг есть оцененные в среднем максимум на 5.0 и минимум на 1.5
- Рекомендуем дополнить библиотеку, добавив книги, вышедшие после 2020 года
- У нас активные пользователи (ставят оценки, пишут обзоры), есть их имена - username - их нужно удерживать, изучать их поведение на ресурсе, для дальнейшего развития сервиса
- Аудитория сервиса скорее всего среднего и молодого возраста, так как количество обзоров и достаточно высокие рейтинги книг у автора книг о Гарри Поттере Джоан Роулинг и саги Стефани Майер "Сумерки"